### PR TITLE
 Add initial AST types & parsing traits

### PR DESCRIPTION
**Pull request by hawkw**
This branch adds an initial set of types representing AST nodes in
the shell's command language. In addition, it defines traits for
parsing strings to AST nodes, and initial parsers for some of the basic
AST types.

### CODE REVIEW

**Comment by hawkw** [comment]
@myrrlyn I'd love a review from you on this branch --- I'd like to get it merged so we can base future changes on it (such as my WIP branch adding rudimentary support for the `cd` builtin).

File: src/ast/builtin.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#[derive(Clone, Debug)]
+pub struct WithEnv<'a> {
+    cmd: ast::Cmd<'a>,
+    // TODO(eliza): what kind of fucked up iterator is the env going to be?
+}
+
+// ===== impl Builtin =====
+
+impl<'a> Parse<'a> for Builtin<'a> {
+    type Error = CdError;
```
Comment (myrrlyn) [inline src/ast/builtin.rs]: This probably shouldn't be limited to only failing on chdir...?

File: src/ast/builtin.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#[derive(Clone, Debug, Fail)]
+pub enum CdError {
+    #[fail(display = "cd: no path provided")]
+    NoPath,
+}
+
+#[derive(Clone, Debug)]
+pub struct WithEnv<'a> {
+    cmd: ast::Cmd<'a>,
+    // TODO(eliza): what kind of fucked up iterator is the env going to be?
```
Comment (myrrlyn) [inline src/ast/builtin.rs]: `Iterator<Item=(&str, &str)>`

File: src/ast/builtin.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+// ===== impl Builtin =====
+
+impl<'a> Parse<'a> for Builtin<'a> {
+    type Error = CdError;
+    fn parse_from(s: &'a str) -> Result<Self, ParseError<Self::Error>> {
+        let mut args = s.trim().split_whitespace();
```
Comment (myrrlyn) [inline src/ast/builtin.rs]: I'll make an iterator of tokens because split_whitespace is incorrect behavior

File: src/ast/builtin.rs (hunk 1 above)
Comment (hawkw) [reply src/ast/builtin.rs]: > This probably shouldn't be limited to only failing on chdir...

Yeah, it won't be when there are more builtins that we can try to parse. Presumably there will be an enum of any special error types that all the builtin types could return. Or, probably, if I actually figured out how to use `failure` correctly, some of the the error type complexity could probably be avoided...

File: src/ast/builtin.rs (hunk 3 above)
Comment (hawkw) [reply src/ast/builtin.rs]: yeah, this should definitely be updated to use your parsers as well --- i probably should've done that, but meh.

**Pull request merged**